Src/FluentAssertions/Equivalency/Steps/EnumerableEquivalencyStep.cs

```diff
@@ -61,10 +61,17 @@ internal static object[] ToArray(object value)
         {
             return ((IEnumerable)value).Cast<object>().ToArray();
         }
-        catch (InvalidOperationException) when (value.GetType().Name.Equals("ImmutableArray`1", StringComparison.Ordinal))
+        catch (InvalidOperationException) when (IsIgnorableArrayLikeType(value))
         {
-            // This is probably a default ImmutableArray<T>
+            // This is probably a default ImmutableArray<T> or an empty ArraySegment.
             return Array.Empty<object>();
         }
     }
+
+    private static bool IsIgnorableArrayLikeType(object value)
+    {
+        return value.GetType() is { } type &&
+                    (type.Name.Equals("ImmutableArray`1", StringComparison.Ordinal) ||
+                     (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(ArraySegment<>)));
```

Comment:
I would prefer either of these two

```cs
private static bool IsIgnorableArrayLikeType1(object value)
{
    var type = value.GetType();
    return type.Name.Equals("ImmutableArray`1", StringComparison.Ordinal) ||
           (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(ArraySegment<>));
}

private static bool IsIgnorableArrayLikeType2(object value)
{
    return value.GetType().Name.Equals("ImmutableArray`1", StringComparison.Ordinal) ||
           (value.GetType().IsGenericType && value.GetType().GetGenericTypeDefinition() == typeof(ArraySegment<>));
}
```

All three variants JITs to the same assembly code

[SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKBuPKQAJgIIAbRgSV04HMA7aNmBsYAQShRsATwAyASwDWMACrSADjAAUEYACsYYDIwBu2NgFcYAShoBvGoyeNiAdlPmrAOgDiMDGqaWtaM8riMdowAvowYGjCMAGSJjs5p6U5acZpeAHLY+DBeAKIAjhbmuFoARJz4+BYYwqISUtIABuTVaIwAyhhQ8vy8AMIQ+OrYg7gQ/F4A8lAAJkPmIYSEqRnbWfFe3H78MINggQnJsXt+GIfH8qfxACIwAGZD8hjys8GMALy/l00EBeWlaMl6MF4hX4GAAPAA+axIgDcNCiNDoDGYrA43D4gikInEkhkCmUZ3IOn0hmMZksNnsWw8UEBCQBdO81zOwVR1DSblZeQKRTKFTYVVq9UazWJbU63T6AyGo3Gk2mswWy1WbHWmz522cuxyBxgRxOZySiUF11u5qer3en2+IX+rOBoJJ0ghUNNcMR1l56No1HoTBY7C4PAEQiJYLkSlU8VIVIMRg89Ns1Ac+qcAo5RS58WCQsKJXKlRqdQaTVjnvlPX6g2GYwmUzCGsWK34a0YGyZO3zvn83Os+1wtvuFoug8LQVHNtNdwemmeb34Hy+/B+ruyMHdce90L9KLRQA=)

Reply:
Addressed in #2511